api/client/client_test.go

```diff
@@ -0,0 +1,214 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/gravitational/teleport/api/client/proto"
+
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/ssh"
+	"google.golang.org/grpc"
+)
+
+// mockServer mocks an Auth Server.
+type mockServer struct {
+	grpc *grpc.Server
+	*proto.UnimplementedAuthServiceServer
+}
+
+func newMockServer() *mockServer {
+	m := &mockServer{
+		grpc.NewServer(),
+		&proto.UnimplementedAuthServiceServer{},
+	}
+	proto.RegisterAuthServiceServer(m.grpc, m)
+	return m
+}
+
+func (m *mockServer) Ping(ctx context.Context, req *proto.PingRequest) (*proto.PingResponse, error) {
+	return &proto.PingResponse{}, nil
+}
+
+// mockInsecureCredentials mocks insecure Client credentials.
+// it returns a nil tlsConfig which allows the client to run in insecure mode.
+type mockInsecureTLSCredentials struct{}
+
+func (mc *mockInsecureTLSCredentials) Dialer(cfg Config) (ContextDialer, error) {
+	return nil, trace.NotImplemented("no dialer")
+}
+
+func (mc *mockInsecureTLSCredentials) TLSConfig() (*tls.Config, error) {
+	return nil, nil
+}
+
+func (mc *mockInsecureTLSCredentials) SSHClientConfig() (*ssh.ClientConfig, error) {
+	return nil, trace.NotImplemented("no ssh config")
+}
+
+func TestNew(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	addr := "localhost:3025"
+	startMockServer(t, addr)
+
+	tests := []struct {
+		desc      string
+		config    Config
+		assertErr require.ErrorAssertionFunc
+	}{{
+		desc: "successfully dial tcp address.",
+		config: Config{
+			Addrs: []string{addr},
+			Credentials: []Credentials{
+				&mockInsecureTLSCredentials{},
+			},
+			DialOpts: []grpc.DialOption{
+				grpc.WithInsecure(),
+			},
+		},
+		assertErr: require.NoError,
+	}, {
+		desc: "synchronously dial addr/cred pairs and succeed with the 1 good pair.",
+		config: Config{
+			Addrs: []string{"bad addr", addr, "bad addr"},
+			Credentials: []Credentials{
+				&tlsConfigCreds{nil},
+				&mockInsecureTLSCredentials{},
+				&tlsConfigCreds{nil},
+			},
+			DialOpts: []grpc.DialOption{
+				grpc.WithInsecure(),
+			},
+		},
+		assertErr: require.NoError,
+	}, {
+		desc: "fail to dial with a bad address.",
+		config: Config{
+			DialTimeout: time.Second,
+			Addrs:       []string{"bad addr"},
+			Credentials: []Credentials{
+				&mockInsecureTLSCredentials{},
+			},
+			DialOpts: []grpc.DialOption{
+				grpc.WithInsecure(),
+			},
+		},
+		assertErr: func(t require.TestingT, err error, _ ...interface{}) {
+			require.Error(t, err)
+			require.Containsf(t, err.Error(), "context deadline exceeded", "")
+		},
+	}}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			t.Parallel()
```

Comment:
Do you need to call Parallel both at the beginning of the test method and in each subtest? I think we usually only call it at the beginning of a test.

Reply:
Yes, this is noted in the [testing rfd](https://github.com/gravitational/teleport/blob/master/rfd/0001-testing-guidelines.md#parallelization), and I tested this to make sure. `t.Parallel` at the top of a `Test` function will run the `Test` function parallel to other `Test` functions. `t.Parallel` within `t.Run` will run the test case parallel with other test cases.